### DISCUSSION TITLE
There's any tool for Keycloak configuration change management?

### DISCUSSION THREAD

**Discussion by Jackojack7** [keycloak/keycloak#19782] [Q&A] [answered]
I have an application deployed on multiple machines that integrates Keycloak as an authentication server. 

Naturally, The Keycloak configurations change over time with each deployment by the client. 

However, I need to periodically make changes to clients, groups, roles, etc. in all Keycloak Realms. 

Is there a known approach to managing these changes to the Keycloak configuration, similar to using Liquibase to update a database? 

I came across a similar [question](https://forum.liquibase.org/t/extension-for-change-management-for-keycloak-rest-configuration-api/6876) on the Liquibase forum from May 22, but I didn't receive any answers. 

Thank you!

**Comment by slominskir** [comment]
You could rely on external user storage providers such as LDAP to manage change management for users, groups, and roles (Keycloak can periodically sync).    Updates to clients would need to be done in Keycloak, but can be scripted using the Admin CLI.

**Reply by darius-m** [reply]
You could also use Ansible with different modules - e.g. https://docs.ansible.com/ansible/latest/collections/community/general/keycloak_realm_module.html to manage various Keycloak settings - realms, clients, roles, etc.. Do note that Ansible usually does not make destructive operations, so probably having a list of realms will just add realms, not remove them.

**Comment by andre-nascimento6791** [comment, marked as answer]
Hello, @slominskir,


Maybe this tool `keycloak-config-cli` could help you:

- https://github.com/adorsys/keycloak-config-cli
- https://hub.docker.com/r/adorsys/keycloak-config-cli
- https://adorsys.com/en/products/keycloak-config-cli/


And I also recommend you take a look at [this useful project ](https://github.com/thomasdarimont/keycloak-project-example)from @thomasdarimont , which uses the aforementioned tool to help in provisioning the Keycloak instances with the needed data loaded.

I hope it helps.

**Reply by slominskir** [reply]
Hi @andre-nascimento-rh,

Keycloak already comes with an admin CLI that allows you to script configuration as code.   That generally requires building bash scripts or similar perhaps with environment variables, but bash and environment variables are ubiquitous and essential for all system admins and developers so they are probably already familiar.  Example:  https://github.com/slominskir/keycloak_cluster/blob/main/keycloak/init/setup.sh.   However, the question above specifically mentioned managing users and groups, which is why I suggested LDAP (as LDAP is the industry standard way to manage users and groups and Keycloak supports syncing to/from).   Keycloak also already provides JSON [import and export](https://www.keycloak.org/server/importExport).

**Comment by dmeyerholt** [comment]
Hi all,
depending on your needs and/or setup this terraform provider could also be of use:
https://registry.terraform.io/providers/mrparkers/keycloak/latest/docs

Best
